blockstore/apps/bundles/data/store.py

```diff
@@ -0,0 +1,222 @@
+"""
+BundleStore is the primary interface for the actual data stored by a Bundle.
+
+This module define the abstract BundleStore interface that backends would need
+to implement, as well as providing definitions for the basic data types that
+BundleStores take as inputs and outputs.
+
+Examples of data structues:
+* An Opaque-Key like object repesenting BundleVersion string representation of
+  UUID + version.
+* Directory listing.
+* File -> URL mapping.
+
+All references to  BundleVersions from this layer should use Bundle UUIDs, and
+not model primary keys.
+"""
+from collections import namedtuple
+import base64
+import hashlib
+import pathlib
+import json
+import uuid
+
+from django.core.files.storage import default_storage
+from django.core.files.base import ContentFile, File
+from django.dispatch import Signal
+
+import attr
+
+
+snapshot_created = Signal(providing_args=["bundle_uuid", "snapshot_hash"])
```

Comment:
Since we are going for immutable data structures (👍), I would recommend that instead of (or at least in addition to) signals, a transaction log is maintained (it could be a simple Django model). This would allow plugins to operate out of process and asynchronously and have far more robust error recovery than possible via patterns like this: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/block_structure/tasks.py#L95-L122.

Reply:
Yeah, transaction log sounds great. I didn't take a stab at it here because there are a few separate-but-related use cases and it's not clear to me what the data model should look like:

**Security audit log.** Permanent storage, search by user or resource. Possibly by time range. Covers changes to mutable data structures (e.g. Collections) as well as Bundle creation. Storage in Django models.

**Who is working on my project now?** Long-asked for Studio feature to show who was recently or who is actively working on a course so they don't step on each others toes. Long term storage probably not necessary, queries more likely at collection level over a limited time range. Might be serviceable using the security audit log model.

**Replay-style transaction logs for plugins.** Verbose, query either the most recent thing or possibly a large chunk. This sounds very much like a fit for Kafka, but Django models and a read-replica might work.

So I agree in principle that we should add this eventually, but I'm not sure exactly what form it should take. I hope that gets clearer as we develop more plugins.